### PR TITLE
Increase the randomness and size of the sessionId (newId()) #1391

### PR DESCRIPTION
**Pull request by MSNev**
The increased size of the newId() value (now length 22 was 5) will change the value reported for the following fields (once all cookies have expired)
- SessionId
- UserId

### CODE REVIEW

File: extensions/applicationinsights-analytics-js/src/JavaScriptSDK/Telemetry/PageViewManager.ts [hunk 1]
```diff
@@ -185,7 +187,7 @@ export class PageViewManager {
                                     pageViewPerformanceSent = true;
                                 }
                             }
-                        } else if (DateTimeUtils.GetDuration(start, +new Date) > maxDurationLimit) {
+                        } else if (start > 0 && DateTimeUtils.GetDuration(start, +new Date) > maxDurationLimit) {
```
Comment (MSNev) [inline extensions/applicationinsights-analytics-js/src/JavaScriptSDK/Telemetry/PageViewManager.ts:190]: Yes, there is a case (mainly Old Hosted IE instances) where the performance.timing values may actually be negative

File: extensions/applicationinsights-analytics-js/src/JavaScriptSDK/Telemetry/PageViewPerformanceManager.ts [hunk 2]
```diff
@@ -163,7 +163,7 @@ export class PageViewPerformanceManager {
         } else {
             // for other page views, don't report if it's outside of a reasonable range
             for (let i = 0; i < durations.length; i++) {
-                if (durations[i] >= this.MAX_DURATION_ALLOWED) {
+                if (durations[i] < 0 || durations[i] >= this.MAX_DURATION_ALLOWED) {
```
Comment (MSNev) [inline extensions/applicationinsights-analytics-js/src/JavaScriptSDK/Telemetry/PageViewPerformanceManager.ts:166]: Yes, there is a case (mainly Old Hosted IE instances) where the performance.timing values may actually be negative

File: extensions/applicationinsights-properties-js/src/Context/Session.ts [hunk 3]
```diff
@@ -74,12 +75,14 @@ export class _SessionManager {
             this.initializeAutomaticSession();
         }
 
-        const now = DateTimeUtils.Now();
+        // Always using Date getTime() as there is a bug in older IE instances that causes the performance timings to have the hi-bit set eg 0x800000000 causing
+        // the number to be incorrect.
+        const now = new Date().getTime();
```
Comment (MSNev) [inline extensions/applicationinsights-properties-js/src/Context/Session.ts:80]: As per above comment just trying to completely avoid the Hosted IE timing issue (DateTimeUtils.Now() tries to use it), and as this is just a calculation for the cookie expiry time we don't need the extra floating point precision anyway -- so just don't use it.

File: extensions/applicationinsights-properties-js/src/Context/Session.ts [hunk 4]
```diff
@@ -166,9 +169,9 @@ export class _SessionManager {
     }
 
     private renew() {
-        const now = DateTimeUtils.Now();
+        const now = new Date().getTime();
```
Comment (MSNev) [inline extensions/applicationinsights-properties-js/src/Context/Session.ts:172]: As above, avoid performance API bugs and we don't need the additional floating point precision.

File: extensions/applicationinsights-properties-js/src/Context/Session.ts [hunk 5]
```diff
@@ -166,9 +169,9 @@ export class _SessionManager {
...
 
-        this.automaticSession.id = Util.newId();
+        this.automaticSession.id = Util.newId((this.config && this.config.idLength) ? this.config.idLength() : 22);
```
Comment (MSNev) [inline extensions/applicationinsights-properties-js/src/Context/Session.ts:174]: Need to add this to the readme.

File: extensions/applicationinsights-properties-js/src/Context/Session.ts [hunk 6]
```diff
@@ -196,15 +199,15 @@ export class _SessionManager {
...
 
         // if sessionExpirationMs is set to 0, it means the expiry is set to 0 for this session cookie
         // A cookie with 0 expiry in the session cookie will never expire for that browser session.  If the browser is closed the cookie expires.  
         // Another browser instance does not inherit this cookie.
         const UTCString = this.config.sessionExpirationMs() === 0 ? '0' : cookieExpiry.toUTCString();
-        Util.setCookie(this._logger, this._storageNamePrefix(), cookie.join('|') + ';expires=' + UTCString, cookieDomnain);
+        Util.setCookie(this._logger, this._storageNamePrefix(), cookie.join('|') + ';expires=' + UTCString, cookieDomain);
 
-        this.cookieUpdatedTimestamp = DateTimeUtils.Now();
+        this.cookieUpdatedTimestamp = new Date().getTime();
```
Comment (MSNev) [inline extensions/applicationinsights-properties-js/src/Context/Session.ts:210]: And again as above comments on performance DateTime issues

File: extensions/applicationinsights-properties-js/src/Context/User.ts [hunk 7]
```diff
@@ -61,7 +61,7 @@ export class User implements IUser {
         this.config = config;
 
         if (!this.id) {
-            this.id = Util.newId();
+            this.id = Util.newId(config.idLength());
```
Comment (MSNev) [inline extensions/applicationinsights-properties-js/src/Context/User.ts:64]: config *should* always exist, but better add null checks

File: shared/AppInsightsCommon/src/Util.ts [hunk 8]
```diff
@@ -479,53 +479,19 @@ export class Util {
...
-        return result;
-    }
+    public static newId = CoreUtils.newId;
 
     /**
      * generate a random 32bit number (-0x80000000..0x7FFFFFFF).
      */
     public static random32() {
-        return (0x100000000 * Math.random()) | 0;
+        return CoreUtils.random32(true);
```
Comment (MSNev) [inline shared/AppInsightsCommon/src/Util.ts:488]: Moved the main implementation of these to CoreUtils, and added the ability to specify if you want signed or unsigned result, CoreUtils.random32(signed?: boolean). As this is a legacy version, this one will always returned a signed result as per documentation.

**Review by markwolff — APPROVED** [review]

**Pull request merged**